docs/proposals-accepted/202304-cardinalitymanagment.md

```diff
@@ -0,0 +1,190 @@
+
+4. We will Implement a new API handler function that
+
+   - Accepts optional query parameters, such as top , for filtering and sorting the results.
+   - Calls the data structure's methods to compute the required statistics.
+   - Serializes the results into JSON format and returns it as an HTTP response.
+   - Handles potential partial results by setting the isPartial field in the response and providing an appropriate error message.
+
+5. (Optional)We can add a flag to the Thanos receiver, allowing users to enable or disable the API.
+
+### Alternatives
+
+#### Two seperate APIs
+
+- We can also perform another API design, where we will have two seperate endpoints for label names cardinality and label values cardinality.Let us take a look at it.
+
+#### Label Names
+
+`GET,POST /api/v1/cardinality/label_names`
+
+This endpoint will be responsible for returning the realtime label names cardinality . It will count distinct label values per label name.
+
+Request parameters:
+
+- selector - (optional) specifies PromQL selector that will be used to filter series that must be analyzed.
+- limit - (optional) specifies max count of items in field cardinality in response (default=20, min=0, max=500)
+
+Response -
+
+```
+{
+"label_values_count_total": <number>,
+"label_names_count": <number>,
+"cardinality": [
+  {
+    "label_name": <string>,
+    "label_values_count": <number>
+  }
+]
+}
+```
+
+#### Label values
+
+This endpoint will return the label values cardinality associated with request parameter `label_names[]`. It will return the series count per label value for each label in the request parameter `label_names[]`.
+
+`GET,POST /api/v1/cardinality/label_values`
+
+Request parameters:
+
+- label_names[] - (required) specifies label names for which label values cardinality will be returned.
+- selector - (optional) specifies PromQL selector that will be used to filter series that must be analyzed.
+- limit - (optional) specifies max count of items in field cardinality in response (default=20, min=0, max=500).
+
+```
+{
+"series_count_total": <number>,
+"labels": [
+  {
+    "label_name": <string>,
+    "label_values_count": <number>,
+    "series_count": <number>,
+    "cardinality": [
+      {
+        "label_value": <string>,
+        "series_count": <number>
+      }
+    ]
+  }
+]
+}
+```
+
+#### Pros
+
+1. It provides more granularity about series data.Also, we can get specific information that we need.Users can choose to query label names, label values, or both depending on their requirements.
+2. It can be more efficient as we query only the data that we need.
+
+### Cons
+
+1. It will be more complex to users as they will have to query two endpoints to get the required information.
+
+### Open Questions
+
+1. Do we need to deduplicate the series before computing the cardinality?
+2. As tenant's data is distributed across multiple receivers, how can we query cardinality information for a specific tenant?
```

Comment:
I think these 2 are something this proposal should address rather than open questions...

Reply:
For point 2, if it is too ambiguous to figure out a way to do it, let's just put it as part of `Non-goals` to reduce the scope. It is totally fine to do that.

Trying to understand point 1 a little bit more, what does it mean by deduplicate the series? Do you mean deduplicate among multiple Receiver replicas? If we want to skip point 2, then we can focus on single receiver first then we don't have to deduplicate. Series on a single receiver instance doesn't have duplication.